### ISSUE TITLE
Add property to constructor

### ISSUE THREAD

**Issue by giggio** [code-cracker/code-cracker#307]
This:

````csharp
class Foo
{
    public int Bar { get; set; }
}
````

Becomes:

````csharp
class Foo
{
    public Foo(int bar)
    {
        this.Bar = bar;
    }
    public int Bar { get; set; }
}
````

If the constructor already exists, add to it:

This:

````csharp
class Foo
{
    public Foo()
    {
    }
    public int Bar { get; set; }
}
````

Becomes:

````csharp
class Foo
{
    public Foo(int bar)
    {
        this.Bar = bar;
    }
    public int Bar { get; set; }
}
````

If more than one constructor exists, offer one fix for each constructor, and another for all constructors.

Diagnostic Id: `CC0083`
Severity: Hidden
Category: Refactoring

**Comment by carloscds** [comment]
@giggio How I know what property can put in constructor ?

Look:
````csharp

class Foo
{
    public int Bar { get; set; }
    public int Some {get; set;}
}
````

**Comment by jwooley** [comment]
I can see this becoming an anti-pattern quickly. What happens if I have 20 properties. I would hate to have a constructor for each property even if they are optional. For me a constructor with that many properties should be refactored. This makes more sense for setting private fields or properties without public setters.

**Comment by giggio** [comment]
@carloscds we would have a diagnostic for each property that isn't already being set on the constructor. On your example, we would have 3 diagnostics:

* Add Bar assignment to constructor
* Add Some to constructor
* Add all properties assignments to constructor

**Comment by giggio** [comment]
@jwooley We would not generate 20 constructor, only one.
We could expand this issue to include private fields. Properties without public setters should already be included.

**Comment by jwooley** [comment]
I was referring to 20 parameters to the single constructor.






Sent from Windows Mail





From: Giovanni Bassi
Sent: ‎Wednesday‎, ‎April‎ ‎8‎, ‎2015 ‎3‎:‎27‎ ‎PM
To: 'code-cracker/code-cracker'
Cc: Jim Wooley





@jwooley We would not generate 20 constructor, only one.
We could expand this issue to include private fields. Properties without public setters should already be included.

**Comment by carloscds** [comment]
@giggio I guess that this issue need more discussion because it's not totally clear.

**Comment by giggio** [comment]
@jwooley This is usually considered a bad smell. So are public setters, put Microsoft still provides the `prop` snippet to help build them. Tools can be abused. CodeCracker suffers the exact same problem any tool would suffer. This is no reason not to do this item, as it is a common pattern and very helpful. And it is a refactoring, so it would only appear when demanded.

**Comment by giggio** [comment]
@carloscds What is not clear? Please post what you did not understand so we can search together for a solution. :)
I marked it as needing more discussion.

**Comment by carloscds** [comment]
@giggio Look at my sample above. I have many properties and fields in my class. What properties/fields I will include as parameter in constructor ? All ? If I do this, is it opposite of issue #44 ?

**Comment by giggio** [comment]
@carloscds We would have 3 code fixes suggested for your example:


* Add Bar assignment to constructor
  * This would only add `Bar`
* Add Some to constructor
  * This would only add `Some`
* Add all properties assignments to constructor
  * This would add bot `Bar` and `Some` to the constructor, that would have 2 parameters.

It is not the opposite of #44 because that creates a **new type** for the already existing parameters. We are not creating a new type. We are simply adding arguments to the constructor.

**Comment by giggio** [comment]
I just figured that this might break code, which is ok, since it is a refactoring. Existing constructor calls would all break. We could add a placeholder to all these calls, which is what R# does, or simply let them break. Which do you prefer?

**Comment by carloscds** [comment]
@giggio Ok, we have some options: 
1. add some property or field according a option menu
2. add all fields
Is it ?

**Comment by giggio** [comment]
@carloscds Yes, that is it.
Right now this issue is considering only properties. We could expand it to use fields or we could create another issue to consider fields.

**Comment by carloscds** [comment]
@giggio Great. I will suggest another thing about paramenter names: instead using lower name, maybe put a prefix in property name. Example:

This:

````csharp
class Foo
{
    public int Bar { get; set; }
}
````

Becomes:

````csharp
class Foo
{
    public Foo(int propBar)
    {
        this.Bar = propBar;
    }
    public int Bar { get; set; }
}
````
What do you think about ?

**Comment by giggio** [comment]
@carloscds We haven't been using prefixes on the CodeCracker projects or any other code fix, so I prefer not to use here as well. This coding style thing is actually something we still have to discuss on this project, but at this moment I would prefer to keep the standard we already are using.

**Comment by carloscds** [comment]
@giggio Ok, in this case maybe we need to put a number after name.

**Comment by giggio** [comment]
@carloscds Why?

**Comment by carloscds** [comment]
@giggio No, forget this. It's a constructor. I can use this.bar = bar.

**Comment by ElemarJR** [comment]
That's really great just for "read-only" objects (with readonly properties)

**Comment by giggio** [comment]
@julianosaless How is this coming? Do you think you will finish until the end of the month?

**Comment by julianosaless** [comment]
yes, i will finish :)

**Comment by giggio** [comment]
Nice, so it goes on alpha6!

**Comment by giggio** [comment]
@julianosaless How is this coming?

**Comment by julianosaless** [comment]
@giggio i'm back to work this  issue

**Comment by giggio** [comment]
Yes! Cool! Remember to fork from the vnext branch, as master now is for stabilization.
